docs/SIL.rst

```diff
@@ -1579,6 +1590,179 @@ of functions returning uninhabited types. An ``unreachable`` instruction that
+~~~~~~~~~~~~~~~~~~~~
+
+Owned
+`````
+
+Owned ownership models "move only" values. We require that each such value is
+consumed exactly once along all program paths. The IR verifier will flag values
+that are not consumed along a path as a leak and any double consumes as
+use-after-frees. We model move operations via "forwarding uses" such as casts
+and transforming terminators (e.x.: switch_enum, checked_cast_br) that transform
+the input value, consuming it in the process, and producing a new transformed
+owned value as a result.
+
+Given any value, one can produce a new "copy" of the owned value via the usage
+of the copy_value instruction. Since in ARC a "copy" of an owned class is just a
+reference count increment, we can use owned values to break down complex ARC
+def-use graphs into individual "move only" values::
+
+  // TODO:
+  // testcase.swift.
+  func doSomething(x : Klass) {
+    let y = x as? OtherKlass
+    switch y {
+    case .none:
+      break
+    case .some(let z):
+      use(z)
+    }
+  }
+
+  // testcase.sil. A possible SILGen lowering
+  sil @doSomething : $@convention(thin) (@guaranteed Klass) -> () {
+  bb0(%0 : @guaranteed Klass):
+    %1 = copy_value %0 : $Klass
+    checked_cast_br %1 : $Klass to $OtherKlass, bbYes, bbNo
+
+  bbYes(%2 : @owned $OtherKlass):
+    %3 = enum $Optional<OtherKlass>, case #Optional.some!enumelt, %2 : $OtherKlass
+    br bbSucc(%3 : $Optional<OtherKlass>)
+
+  bbNo(%3 : @owned $Klass):
+    destroy_value %3 : $Klass
+    %4 = enum $Optional<OtherKlass>, case #Optional.none!enumelt
+    br bbSucc(%4 : $Optional<OtherKlass>)
+
+  bbSucc(%5 : @owned $Optional<OtherKlass>):
+    switch_enum %5 : $Optional<OtherKlass>, case #Optional.some!enumelt: bbSome, case #Optional.none!enumelt: bbNone
+
+  bbSome(%6 : @owned $OtherKlass):
+    apply %nonConsumingUseOfOtherKlass(%6) : ...
+    destroy_value %6 : $OtherKlass
+    br bbEpilog
+
+  bbNone:
+    br bbEpilog
+
+  bbEpilog:
+    %9999 = tuple()
+    return %9999 : $()
+  }
+
+Notice how our individual copy threads its way through the IR. At any single
+program point, the object is only bound to a single of our owned values.
+
+Guaranteed
+``````````
+
+Guaranteed ownership models values that have a scoped dependent lifetime on a
+base value with owned or guaranteed ownership. These scopes are introduced into
+the IR by explicit begin/end scope instructions. These instructions summarize
+explicitly the region of code where the guaranteed value can be used resulting
+in simplified analysis and verification. A base value must be statically live
+until the end of the scoped region. This ensure that our base value is never
+invalidated while our guaranteed value is live. Values with guaranteed ownership
+follow a dataflow rule that states that any "forwarding" uses of the guaranteed
+value are also guaranteed and are recursively validated as being in the original
+values scope. This was a choice we made to reduce idempotent scopes in the IR.
```

Comment:
It's maybe worth mentioning here that the purpose of "end_borrow" is to tell optimizations to not shrink the base value's life range across this point.